### ISSUE TITLE
CRI: instruct kubelet to (optionally) consume container stats from CRI

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#46984]
#45614 augmented CRI with container stats method. 
 - [ ] Make cadvisor (`pkg/kubelet/cadvisor`) implement the container stats part of the summary api (`pkg/kubelet/server/stats`).
 - [ ] Refactor kubelet to have the flexibility to consume container stats either from cadvisor or from CRI.

We may also need to modify cadvsior to report UUID for the as the storage identifier.

Related issues:
 - ImageFS implementation (#33048)
 - Previous meta issue: #27097. 
/cc @dchen1107 @feiskyer @Random-Liu

**Comment by yujuhong** [comment]
@mrunalp, this is needed for cri-o (or any other cri integration) to support container stats, and will involve changes to mostly kubelet code and a little bit of cadvisor code. If anyone is willing to help out, that'd be great.

**Comment by WIZARD-CXY** [comment]
@yujuhong I think I can give a hand because I have experience on cadvisor and kubelet before.

**Comment by feiskyer** [comment]
@WIZARD-CXY Cool. Thanks.

**Comment by feiskyer** [comment]
@WIZARD-CXY Are you still working on this feature?

**Comment by yujuhong** [comment]
Sorry I forgot to update the issue. @yguo0905 agreed to help with this issue when we talked about Q3 planning a while ago. I assigned the issue to him.

@WIZARD-CXY, if you are still interested in working on this, could you make sure to talk to @yguo0905 before implementing? Thanks!

**Comment by WIZARD-CXY** [comment]
Cool, I will talk to @yguo0905  before implementing.

**Comment by WIZARD-CXY** [comment]
After doing some homework, I know stats.Summary has two parts: one is machine stat and the other is pod stats.The pod stats provider used for now is kubelet. It calls GetContainerInfoV2 to get pod stats
```
func (kl *Kubelet) GetContainerInfoV2(name string, options cadvisorapiv2.RequestOptions) (map[string]cadvisorapiv2.ContainerInfo, error) {
	return kl.cadvisor.ContainerInfoV2(name, options)
}
```
I think we can extend this function like below to fullfill "Refactor kubelet to have the flexibility to consume container stats either from cadvisor or from CRI." as yujuhong's need.
```
   if using cri {
        // get stats from cri
  } else{
      // get stats from cadvisor
  }

```
@yguo0905  what's your opinion on this?

**Comment by WIZARD-CXY** [comment]
@yujuhong what about your need "Make cadvisor (pkg/kubelet/cadvisor) implement the container stats part of the summary api (pkg/kubelet/server/stats)." ?
I think the cadvisor has already implemented the summary api. see this func:
```
func (kl *Kubelet) GetContainerInfoV2(name string, options cadvisorapiv2.RequestOptions) (map[string]cadvisorapiv2.ContainerInfo, error) {
	return kl.cadvisor.ContainerInfoV2(name, options)
}
```

So I don't quite understand what you want to accomplish?

**Comment by yguo0905** [comment]
> @yguo0905 what's your opinion on this?

Yeah, the idea is correct. :) I talked to @yujuhong the other day. The `StatsProvider` interface used by the summary API uses cadvisor types, so we need to modify this interface to accommodate both the cadvisor and the CRI metrics for container stats. I will come up with a plan some time this week.

**Comment by WIZARD-CXY** [comment]
OK

**Comment by WIZARD-CXY** [comment]
@yujuhong any update on this?

**Comment by yguo0905** [comment]
The [`StatsProvider`](https://github.com/kubernetes/kubernetes/blob/b32639f9e7d47d0f69bb41b5535803ae4c9d835a/pkg/kubelet/server/stats/handler.go#L41) interface can be changed to accommodate both the cadvisor and the CRI stats, as below.

```
type StatsProvider interface {
    // Interfaces that will be implemented by both cadvisor and CRI. Provides containers stats for all pods and imageFS stats.
    ListPodStats() ([]statsapi.PodStats, error)
    ImageFsInfo() (*statsapi.FsStats, error)

    // Interfaces that will be implemented by cadvisor only. Provides cgroup stats and rootFS stats.
    GetContainerStats(containerName string) (*statsapi.ContainerStats, *statsapi.NetworkStats, error)
    RootFsInfo() (*statsapi.FsStats, error)

    // Provides the stats that are neutral to cadvisor vs CRI.
    GetNode() (*v1.Node, error)
    GetNodeConfig() cm.NodeConfig
    GetPods() []*v1.Pod
    GetPodByName(namespace, name string) (*v1.Pod, bool)
    ListVolumesForPod(podUID types.UID) (map[string]volume.Volume, bool)
}
```

- The interface will use summary API's types (`statsapi`)
- The conversion between the cadvisor/CRI types to the summary API's types will be moved to `kubelet_cadvisor.go` (and probably rename it to something like `kubelet_summary_stats.go`).

In addition, the following tables describe the data source of the salient fields in the summary API, in both `cadvisor` and `cri` mode.
 
| NodeStats | when using cadvisor | when using CRI |
| ------ | ------ | ------ |
| `Node.NodeName` | from kubelet | (same)
| `Node.SystemContainers` | cgroup stats from cadvisor | (same)
| `Node.CPU` | root cgroup stats from cadvisor | (same)
| `Node.Memory` | root cgroup stats from cadvisor | (same)
| `Node.Network` | root cgroup stats from cadvisor | (same)
| `Node.Fs` | `cadvisor.RootFsInfo()` | (same)
| `Node.Runtime.ImageFs` | `cadvisor.ImagesFsInfo()` | `ImageManagerService.ImageFsInfo()`

| PodStats | when using cadvisor | when using CRI |
| ------ | ------ | ------ |
| `Pods.PodRef` | from container label | `PodSandboxManager.PodSandboxStatus()`
| `Pods.Containers.CPU` | cgroup stats from cadvisor | `ContainerStatsManager.ListContainerStats()`
| `Pods.Containers.Memory` | cgroup stats from cadvisor| `ContainerStatsManager.ListContainerStats()`
| `Pods.Containers.Rootfs.[UsedBytes\|InodesUsed]` | cgroup stats from cadvisor | `ContainerStatsManager.ListContainerStats()`
| `Pods.Containers.Rootfs.(others)` | `cadvisor.ImagesFsInfo()` | `ImageManagerService.ImageFsInfo()`
| `Pods.Containers.Logs.[UsedBytes\|InodesUsed]` | cgroup stats from cadvisor | (not available)
| `Pods.Containers.Logs.(others)` | `cadvisor.RootFsInfo()` | (same)
| `Pods.Containers.UserDefinedMetrics` | cgroup stats from cadvisor | (not available)
| `Pods.Network` | cgroup stats from cadvisor | (not available)
| `Pods.VolumeStats` | from kubelet | (same)

**Comment by WIZARD-CXY** [comment]
let me have a closer look

**Comment by yguo0905** [comment]
Cool. BTW, I've sent you a message on slack.
